### PR TITLE
CI (Buildkite): Run the Distributed test suite with multithreading enabled

### PR DESCRIPTION
**Pull request by DilumAluthge**

### CODE REVIEW

**Comment by DilumAluthge** [comment]
@vchuravy Would it be feasible to make the Distributed test suite thread-safe?

**Comment by DilumAluthge** [comment]
The specific test failure is:

```
Distributed                         (1) |        started at 2021-10-03T03:41:05.590

Test Failed at /cache/build/amdci7-2/julialang/julia-master/julia-bccbf7b7cd/share/julia/stdlib/v1.8/Distributed/test/distributed_exec.jl:146

  Expression: remotecall_fetch((k->begin

                yield()

                haskey(Distributed.PGRP.refs, k)

            end), id, fid) == false

   Evaluated: true == false

ERROR: LoadError: There was an error during testing

in expression starting at /cache/build/amdci7-2/julialang/julia-master/julia-bccbf7b7cd/share/julia/stdlib/v1.8/Distributed/test/distributed_exec.jl:160

Distributed                         (1) |         failed at 2021-10-03T03:41:22.950
```

File: stdlib/Distributed/test/distributed_exec.jl [hunk 1]
```diff
@@ -143,8 +143,11 @@ function test_futures_dgc(id)
     @test fetch(f) == id
     @test f.v !== nothing
     yield(); # flush gc msgs
-    @test remotecall_fetch(k->(yield();haskey(Distributed.PGRP.refs, k)), id, fid) == false
-
+    if Threads.nthreads() > 1
+        @test_broken remotecall_fetch(k->(yield();haskey(Distributed.PGRP.refs, k)), id, fid) == false
```
Comment (vchuravy) [inline stdlib/Distributed/test/distributed_exec.jl:147]: No this won't help. Any I am not sure that this test even makes sense...

File: stdlib/Distributed/test/distributed_exec.jl [hunk 2]
```diff
@@ -143,9 +143,8 @@ function test_futures_dgc(id)
     @test fetch(f) == id
     @test f.v !== nothing
     yield(); # flush gc msgs
```
Comment (vchuravy) [inline stdlib/Distributed/test/distributed_exec.jl:145]: I believe this yield is not correct, same with the yield on the other side.
Nothing guarantees that the flush gc_msgs task has run. 

Same with the yield on the other side.
@vtjnash any ideas on redeeming this test, short of yanking it out?

Comment (vtjnash) [reply stdlib/Distributed/test/distributed_exec.jl:145]: Probably can't do much better than loop the test below for a couple seconds, and see if it finishes within that time

**Comment by DilumAluthge** [comment]
With the following patch, the Distributed tests pass for me locally when multithreading is enabled. I've added that patch to this PR; let's see if the Distributed tests pass with multithreading on CI.

```patch
diff --git a/stdlib/Distributed/test/distributed_exec.jl b/stdlib/Distributed/test/distributed_exec.jl
index 32b6b703f3..4063114df3 100644
--- a/stdlib/Distributed/test/distributed_exec.jl
+++ b/stdlib/Distributed/test/distributed_exec.jl
@@ -143,6 +143,8 @@ function test_futures_dgc(id)
     @test fetch(f) == id
     @test f.v !== nothing
     yield(); # flush gc msgs
+    sleep(6);
+    yield();
     @test remotecall_fetch(k->(yield();haskey(Distributed.PGRP.refs, k)), id, fid) == false


@@ -153,6 +155,8 @@ function test_futures_dgc(id)
     @test f.v === nothing
     finalize(f)
     yield(); # flush gc msgs
+    sleep(6);
+    yield();
     @test remotecall_fetch(k->(yield();haskey(Distributed.PGRP.refs, k)), id, fid) == false
 end

@@ -243,6 +247,8 @@ function test_remoteref_dgc(id)
     @test remotecall_fetch(k->(yield();haskey(Distributed.PGRP.refs, k)), id, rrid) == true
     finalize(rr)
     yield(); # flush gc msgs
+    sleep(6);
+    yield();
     @test remotecall_fetch(k->(yield();haskey(Distributed.PGRP.refs, k)), id, rrid) == false
 end
 test_remoteref_dgc(id_me)
```

**Comment by krynju** [comment]
Passes locally in multithreaded on this branch https://github.com/JuliaLang/julia/pull/42339
Not on master though

Main worker with these threads is involved in that testing, so seems like there's some race on that cleanup

**Comment by tkf** [comment]
If you skim Distributed.jl, you can spot a lot of data races (which may or may not be relevant in practice). I wonder if it makes sense to use the experimental pipeline?

**Comment by vchuravy** [comment]
Yeah making Distributed.jl is an ongoing project and an aspiration

File: stdlib/Distributed/test/distributed_exec.jl (hunk 2 above)
Comment (tkf) [reply stdlib/Distributed/test/distributed_exec.jl:145]: @DilumAluthge I think what Jameson was saying was something like

```julia
@test begin
    t0 = time_ns()
    while remotecall_fetch(k->(yield();haskey(Distributed.PGRP.refs, k)), id, fid) != false
        sleep(0.1)
        if (time_ns() - t0) / 1e9 > 10  # wait 10s
            error("timeout")
        end
    end
    true
end
```

Comment (DilumAluthge) [reply stdlib/Distributed/test/distributed_exec.jl:157]: Ahhh, I see. I've pushed a new commit that implements your code. I went ahead and did a 60 second timeout.

**Comment by DilumAluthge** [comment]
I've made a separate PR with the "loop the test for a couple seconds" suggestion from @vtjnash, with the code from @tkf.

https://github.com/JuliaLang/julia/pull/42499

**Comment by DilumAluthge** [comment]
I've been backporting all of the CI PRs to 1.6 and 1.7, because it's the easiest and fastest way to get Buildkite set up across 1.6, 1.7, and master. And if we backport all the PRs, everything backports automatically and cleanly, and there's no need for manual intervention.

Eventually, we will stop backporting CI PRs, and we will start maintaining separate Buildkite configurations on each of the release branches. But I'd like to delay that as much as possible.

But, most importantly, we're not going to start running the Distributed test suite with multithreading on 1.6 or 1.7. We'll only do so starting with 1.8.

The relevant bit from the Buildkite config:

```yaml
if [[ "$${BUILDKITE_PIPELINE_SLUG:?}"   == "julia-release-1-dot-6" ]]; then
  # On Julia 1.6, the Distributed test suite is expected to fail when multithreading is enabled.
  $${JULIA_BINARY:?} -e 'Base.runtests(["all", "--skip", "Distributed"]; ncores = Sys.CPU_THREADS)'
elif [[ "$${BUILDKITE_PIPELINE_SLUG:?}" == "julia-release-1-dot-7" ]]; then
  # On Julia 1.7, the Distributed test suite is expected to fail when multithreading is enabled.
  $${JULIA_BINARY:?} -e 'Base.runtests(["all", "--skip", "Distributed"]; ncores = Sys.CPU_THREADS)'
else
  $${JULIA_BINARY:?} -e 'Base.runtests(["all"]; ncores = Sys.CPU_THREADS)'
fi
```

**Pull request merged**